_posts/2023-01-16-native-adopts-adaptive-gc-policy.adoc

```diff
@@ -0,0 +1,150 @@
+
+At this point we thought,
+do the assumptions made about the default GC policy still apply in 2022?
+So, we removed the GC policy configuration tweak,
+repeated the test and we observed the following memory consumption:
+
+image::adaptive-memory-consumption.png[]
+
+For the same workload the default GC policy, called adaptive,
+would seem to consume close to 50% less heap compared to the space/time.
+Obviously, these graphs alone are not enough to make the switch since we could have a situation where adaptive is using less memory because the overall throughput is less.
+So, let’s look at the details of the benchmark that generated the graphs above and see what throughput numbers we obtain.
+Using https://github.com/Hyperfoil/Hyperfoil[Hyperfoil],
+space/time policy reported these numbers on our environment:
+
+[source,bash]
+----
+[hyperfoil@in-vm]$ wrk -t 128 -c 512 -H 'accept: text/plain' -d 16m http://<host>:8080/hello
+PHASE        METRIC   THROUGHPUT    REQUESTS  ... TIMEOUTS  ERRORS  BLOCKED   2xx
+test         request  93.79k req/s  90036541  ...        0       0      0 ns  90036094
+----
+
+And here are the numbers for adaptive:
+
+[source,bash]
+----
+[hyperfoil@in-vm]$ wrk -t 128 -c 512 -H 'accept: text/plain' -d 16m http://<host>:8080/hello
+PHASE        METRIC   THROUGHPUT    REQUESTS  ... TIMEOUTS  ERRORS  BLOCKED   2xx
+test         request  93.05k req/s  89329151  ...         0       0     0 ns  89328711
+----
+
+NOTE: The results were obtained with `wrk`,
+which is known to have issues with latency numbers
+(see https://redhatperf.github.io/post/coordinated-omission[this blog post] for more details),
+so we can ignore those in the context of this blog post and focus on throughput numbers.
+
+For the same workload,
+adaptive throughput is within 1% of the one obtained with space/time.
+So getting pretty much the same throughput with adaptive as with space/time and close to 50% less memory consumption,
+made it a pretty convincing argument to switch to adaptive as default GC policy for Quarkus,
+as it was already the case for other GraalVM.
+
+IMPORTANT: The memory consumption benefits do not apply evenly across all heap sizes.
+Numbers like the ones published in this blog post would apply for maximum heap sizes that are equal or above 3GB,
+at which stage the default minimum heap size is set to ~512MB unless configured otherwise.
+For smaller maximum heap sizes, the memory consumption improvements might be smaller or non-existent.
+
+We often see tests or benchmarks run with no `-Xmx` configured,
+in which case as stated above,
+the maximum heap size is set to 80% of the available physical memory and this heap size can easily exceed 3GB in modern hardware.
+These users would see better out of the box experience with adaptive GC policy.
+
+So, starting with Quarkus 2.14.Final, the GC policy in Quarkus was aligned with GraalVM to be adaptive.
+Even with this change,
+it is still possible to set the GC policy back to space/time.
```

Comment:
```suggestion
It is still possible to set the GC policy back to space/time, should that work better in a specific case.
```

Reply:
Also consider `space/time` -> `"by space and time"` for consistency.